src/controllers/hid/hidcontroller.cpp

```diff
@@ -277,6 +277,38 @@ void HidController::send(QByteArray data, unsigned int reportID) {
     data.prepend(reportID);
 
     int result = hid_write(m_pHidDevice, (unsigned char*)data.constData(), data.size());
+    if (result == -1) {
+        if (ControllerDebug::enabled()) {
+            qWarning() << "Unable to send data to" << getName()
+                       << "serial #" << hid_serial << ":"
+                       << safeDecodeWideString(hid_error(m_pHidDevice), 512);
+        } else {
+            qWarning() << "Unable to send data to" << getName() << ":"
+                       << safeDecodeWideString(hid_error(m_pHidDevice), 512);
+        }
+    } else {
+        controllerDebug(result << "bytes sent to" << getName()
+                               << "serial #" << hid_serial
+                               << "(including report ID of" << reportID << ")");
+    }
+}
+
+void HidController::send_feature_report(
+        QList<int> data, unsigned int length, unsigned int reportID) {
+    Q_UNUSED(length);
+    QByteArray temp;
+    foreach (int datum, data) {
+        temp.append(datum);
+    }
+    send_feature_report(temp, reportID);
+}
+
+void HidController::send_feature_report(QByteArray data, unsigned int reportID) {
+    // Append the Report ID to the beginning of data[] per the API..
+    data.prepend(reportID);
+
+    int result = hid_send_feature_report(
+            m_pHidDevice, (unsigned char*)data.constData(), data.size());
```

Comment:
```suggestion
            m_pHidDevice, static_cast<unsigned char*>data.constData(), data.size());
```

Reply:
I tried this, but I got the following compiler failure:
[CXX] src\controllers\hid\hidcontroller.cpp
hidcontroller.cpp
src\controllers\hid\hidcontroller.cpp(312): error C2061: Syntaxfehler: Bezeichner "dataArray"
src\controllers\hid\hidcontroller.cpp(312): error C2146: Syntaxfehler: Fehlendes "(" vor Bezeichner "dataArray"
src\controllers\hid\hidcontroller.cpp(311): error C2660: "hid_send_feature_report": Funktion akzeptiert keine 1 Argumente
lib\hidapi\hidapi\hidapi.h(296): note: Siehe Deklaration von "hid_send_feature_report"
src\controllers\hid\hidcontroller.cpp(312): error C3927: "->": Ein nachstehender Rückgabetyp ist nach einem Nichtfunktionsdeklarator nicht zulässig.
src\controllers\hid\hidcontroller.cpp(312): error C3484: Syntaxfehler: "->" wird vor dem Rückgabetyp erwartet.
src\controllers\hid\hidcontroller.cpp(312): error C3613: Rückgabetyp nach '"->" (Annahme: "int") fehlt
src\controllers\hid\hidcontroller.cpp(312): error C2371: "dataArray": Neudefinition; unterschiedliche Basistypen
src\controllers\hid\hidcontroller.cpp(303): note: Siehe Deklaration von "dataArray"
src\controllers\hid\hidcontroller.cpp(312): error C2146: Syntaxfehler: Fehlendes ";" vor Bezeichner "size"
src\controllers\hid\hidcontroller.cpp(312): error C3861: "size": Bezeichner wurde nicht gefunden.
src\controllers\hid\hidcontroller.cpp(312): error C2059: Syntaxfehler: ")"
scons: *** [win64_build\src\controllers\hid\hidcontroller.obj] Error 2

Reply:
Forgot the braces after `>` :D

Reply:
Only a opening brace? Isn't a closing brace needed anywhere too?

Reply:
Yes, after `data.constData()`.

Reply:
This does not work:
[CXX] src\controllers\hid\hidcontroller.cpp
hidcontroller.cpp
src\controllers\hid\hidcontroller.cpp(312): error C2440: "static_cast": "const char *" kann nicht in "unsigned char *" konvertiert werden
src\controllers\hid\hidcontroller.cpp(312): note: Die Typen, auf die verwiesen wird, sind nicht verknüpft; die Konvertierung erfordert einen reinterpret_cast-Op
erator oder eine Typumwandlung im C- oder Funktionsformat.
src\controllers\hid\hidcontroller.cpp(311): error C2660: "hid_send_feature_report": Funktion akzeptiert keine 2 Argumente
lib\hidapi\hidapi\hidapi.h(296): note: Siehe Deklaration von "hid_send_feature_report"
scons: *** [win64_build\src\controllers\hid\hidcontroller.obj] Error 2
scons: building terminated because of errors.

Why not keep the proven code from the .send method?

Reply:
We shouldn't use c-style casts for that exact reason. It hides that these type are incompatible and this cast actually reinterprets that data. We need reinterpret_cast here.

Reply:
reinterpret_cast<const unsigned char*>(dataArray.constData()) works